### PR TITLE
Added OutsideClickHandler component with onOutsideClick callback

### PR DESCRIPTION
**Pull request by mattRoyten**
Corrected onOutsideClick functionality missing from the DayPickerSingleDateController component. This prop was within the defaultProps and now is properly applied.

### CODE REVIEW

File: src/components/DayPickerSingleDateController.jsx [hunk 1]
```diff
@@ -548,39 +549,44 @@ export default class DayPickerSingleDateController extends React.Component {
...
-        renderCalendarInfo={renderCalendarInfo}
-        isFocused={isFocused}
-        getFirstFocusableDay={this.getFirstFocusableDay}
-        onBlur={this.onDayPickerBlur}
-        phrases={phrases}
-        daySize={daySize}
-        isRTL={isRTL}
-      />
+      <OutsideClickHandler
+        onOutsideClick={onOutsideClick}
```
Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: if `onOutsideClick` is omitted, should the DayPicker still be wrapped?

Comment (mattRoyten) [inline src/components/DayPickerSingleDateController.jsx]: I do not see any negative impact if `onOutsideClick` is omitted. Are you thinking to conditionally wrap the DayPicker component depending on the `onOutsideClick` value?

Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: Yes, that way if people don't want outside click behavior, they wouldn't incur any overhead.

**Comment by mattRoyten** [comment]
@ljharb - Updated PR with suggestion.

File: src/components/DayPickerSingleDateController.jsx [hunk 2]
```diff
@@ -548,39 +549,47 @@ export default class DayPickerSingleDateController extends React.Component {
       isFocused,
       isRTL,
       phrases,
+      onOutsideClick,
     } = this.props;
 
     const { currentMonth, visibleDays } = this.state;
 
+    const DayPickerWrapper = (onOutsideClick) ? OutsideClickHandler : 'div';
```
Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: Let's remove these unnecessary parents on the ternary conditionals.

File: src/components/DayPickerSingleDateController.jsx [hunk 3]
```diff
@@ -548,39 +549,47 @@ export default class DayPickerSingleDateController extends React.Component {
...
+    const wrapperProps = (onOutsideClick) ? { onOutsideClick } : {};
```
Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: This shouldn't be needed; `onOutsideClick={null}` on a DOM element is a no-op.

**Comment by majapw** [comment]
I'm interested in why the refactor failed. I too think I would prefer what @ljharb suggested and don't know why that would cause tests to fail.

**Comment by mattRoyten** [comment]
@majapw - all of the DayPickerSingleDateController tests were getting the following error:
_Invariant Violation: Unexpected node: [object Object]_

**Comment by ljharb** [comment]
@mattRoyten You had an error in your code, in the unwrapped case, you were returning an object (ie, `{ dayPickerComponent }` instead of directly returning `dayPickerComponent`.

Could you rebase on the command line so as to remove all merge commits, and also remove the last revert commit, and fix that error?

**Comment by mattRoyten** [comment]
@ljharb - got it. Updated.

File: src/components/DayPickerSingleDateController.jsx [hunk 4]
```diff
@@ -565,40 +566,51 @@ export default class DayPickerSingleDateController extends React.Component {
...
-        renderCalendarInfo={renderCalendarInfo}
-        isFocused={isFocused}
-        getFirstFocusableDay={this.getFirstFocusableDay}
-        onBlur={this.onDayPickerBlur}
-        phrases={phrases}
-        daySize={daySize}
-        isRTL={isRTL}
-      />
-    );
+    const dayPickerComponent = (<DayPicker
```
Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: one last nit: the `(` should end the line, and the `);` should begin one - iow:
```jsx
const foo = (
  <div>
  </div>
);
```

File: src/components/DayPickerSingleDateController.jsx [hunk 5]
```diff
@@ -599,6 +601,18 @@ export default class DayPickerSingleDateController extends React.Component {
         isRTL={isRTL}
       />
     );
+
+    if (onOutsideClick) {
```
Comment (majapw) [inline src/components/DayPickerSingleDateController.jsx]: Oh this looks *so* much better

**Comment by mattRoyten** [comment]
@majapw - tests added!

File: test/components/DayPickerSingleDateController_spec.jsx [hunk 6]
```diff
@@ -1401,4 +1402,22 @@ describe('DayPickerSingleDateController', () => {
...
   });
+
+  describe('onOutsideClick', () => {
+    it('should render OutsideClickHandler as it has onOutsideClick prop', () => {
+      const wrapper = shallow(
+        <DayPickerSingleDateController
+          onOutsideClick={() => null}
+        />,
+      );
+      expect(wrapper.find(OutsideClickHandler).length).to.equal(1);
```
Comment (ljharb) [inline test/components/DayPickerSingleDateController_spec.jsx]: `expect(wrapper.find(OutsideClickHandler).to.have.lengthOf(1)` has a nicer failure message; nbd tho :-)

Comment (mattRoyten) [inline test/components/DayPickerSingleDateController_spec.jsx]: sure, want me to change this as well?

Comment (ljharb) [inline test/components/DayPickerSingleDateController_spec.jsx]: Sure, that'd be great :-D

Comment (mattRoyten) [inline test/components/DayPickerSingleDateController_spec.jsx]: done! 💥

**Pull request merged**